### DISCUSSION TITLE
Conditional rendering lit element in react loop

### DISCUSSION THREAD

**Discussion by Satyam** [lit/lit#3123] [Q&A] [answered]
I have the following code to render various icons

```ts
@customElement('icon-check-false')
export class IconCheckFalse extends IconBase {
  svg = svgSquare;
}

@customElement('icon-check-true')
export class IconCheckTrue extends IconBase {
  svg = svgCheckSquareFill;
}

@customElement('icon-check')
export class IconCheck extends IconBase {
  @property({
    type: Boolean,
    // converter is required because default converter simply checks the presence of the attribute
    // not the value so value={false} is still true
    converter: (value, type) => {
      console.log('converter', value, type);
      return value;
    },
  })
  value = false;

  svg = this.value ? svgCheckSquareFill : svgSquare;
}
```
Where `IconBase` is:
```ts
export abstract class IconBase extends LitElement {
  protected abstract svg: TemplateResult<2>;
  override render() {
    return html`<svg
        xmlns="http://www.w3.org/2000/svg"
        width="16"
        height="16"
        fill="currentColor"
        viewBox="0 0 16 16"
      >
        ${this.svg}
      </svg>
      <slot></slot>`;
  }
}
```
`IconBase`  works fine for plenty of other icons which I use all over the React application.  However, `<icon-check>` does not work. 

I tried to use `<icon-check>` to represent a boolean value in a table and in all attempts the whole table shows the same icon, either checked or unchecked, for all rows, as if the icon was cached somehow or lit could not detect the change in what I tried to render.

As you can see, I show the value being converted via a `console.log`, which clearly shows the value reaches the element.  However, that is not reflected in the UI.

I tried several ways of rendering it, for example, not using `IconBase` and conditionally rendering either `<icon-check-true>` or `<icon-check-false>` , for example:

```ts
@customElement('icon-check')
export class IconCheck extends LitElement {
  @property({
    type: Boolean,
    // converter is required because default converter simply checks the presence of the attribute
    // not the value so value={false} is still true
    converter: (value, type) => {
      console.log('converter', value, type);
      return value;
    },
  })
  value = false;

  override render() {
    return this.value
      ? html`<icon-check-true></icon-check-true>`
      : html`<icon-check-false></icon-check-false>`;
  }
/* Also:
  override render() {
    if (this.value) return html`<icon-check-true></icon-check-true>`;
    return html`<icon-check-false></icon-check-false>`;
  }
*/
}
```
Or even following the advice of https://github.com/lit/lit/issues/942 further enclosing the whole conditional in a template.

In the end, I opted to place the condition in the React code, and it works fine, as expected  Note that I am using the same elements derived from `IconBase`, I am just placing the conditional in the React code instead of Lit:
```html
<td className="text-center">
  {venta.iva ? (
    <icon-check-true></icon-check-true>
  ) : (
    <icon-check-false></icon-check-false>
  )}
</td>
```
This solves my immediate problem, but I still wonder what am I doing wrong.

Thanks in advance for any help.

**Comment by augustjk** [comment, marked as answer]
```ts
@customElement('icon-check')
export class IconCheck extends IconBase {
  @property({
    type: Boolean,
    // converter is required because default converter simply checks the presence of the attribute
    // not the value so value={false} is still true
    converter: (value, type) => {
      console.log('converter', value, type);
      return value;
    },
  })
  value = false;

  svg = this.value ? svgCheckSquareFill : svgSquare;
}
```
This original approach would not work since `svg` is only assigned once and `this.value` is not read from again.

The conditional rendering approach is in the right direction but the attribute converter for the boolean property needs tweaking.
```ts
@customElement('icon-check')
export class IconCheck extends LitElement {
  @property({
    type: Boolean,
    converter: (value, type) => {
      return value === 'true';
    },
  })
  value = false;

  override render() {
    return this.value
      ? html`<icon-check-true></icon-check-true>`
      : html`<icon-check-false></icon-check-false>`;
  }
}
```
Within the converter, `value` is of type string (since it's coming from the attribute), so we need to convert this to a boolean.

Here's an example https://lit.dev/playground/#gist=149ffe1726ef0256302dec54a4f2e326

**Reply by Satyam** [reply]
Thanks @augustjk !   I was completely blinded by my React instinct assuming that if I called it from React like this:
```
<icon-check value={!!venta.iva}></icon-check>
```
the boolean nature of the value would be preserved but, of course, that is only valid within React, it doesn't apply once out of it.  And, I was reinforced in my mistake by the printout of `console.log(value, type)` which, of course, always showed the `type` was `boolean` totally blinded to the fact that the `type` reported was the type **expected** not the `typeof value`.   

Thanks again.
